readthedocs/projects/models.py

```diff
@@ -541,13 +555,117 @@ def get_production_media_url(self, type_, version_slug):
+            path = f'//{domain}/{self.proxied_api_url}downloads/{self.language}/{version_slug}/{type_}/'  # noqa
 
         return path
 
+    @property
+    def proxied_api_host(self):
+        """
+        Used for the proxied_api_host in javascript.
+
+        This needs to start with a slash at the root of the domain,
+        and end without a slash
+        """
+        if self.urlconf:
+            # Add our proxied api host at the first place we have a $variable
+            # This supports both subpaths & normal root hosting
+            url_prefix = self.urlconf.split('$', 1)[0]
+            return '/' + url_prefix.strip('/') + '/_'
+        return '/_'
+
+    @property
+    def proxied_api_url(self):
+        """
+        Like the api_host but for use as a URL prefix.
+
+        It can't start with a /, but has to end with one.
+        """
+        return self.proxied_api_host.strip('/') + '/'
+
+    @property
+    def real_urlconf(self):
+        """
+        Convert User's URLConf into a proper django URLConf.
+
+        This replaces the user-facing syntax with the regex syntax.
+        """
+        to_convert = self.urlconf
+
+        # We should standardize these names so we can loop over them easier
+        to_convert = to_convert.replace(
+            '$version',
+            '(?P<version_slug>%s)' % pattern_opts['version_slug']
+        )
+        to_convert = to_convert.replace(
+            '$language',
+            '(?P<lang_slug>%s)' % pattern_opts['lang_slug']
+        )
+        to_convert = to_convert.replace(
+            '$filename',
+            '(?P<filename>%s)' % pattern_opts['filename_slug']
+        )
+        to_convert = to_convert.replace(
+            '$subproject',
+            '(?P<subproject_slug>%s)' % pattern_opts['project_slug']
+        )
+
+        if '$' in to_convert:
+            log.warning(
+                'Unconverted variable in a project URLConf: project=%s, to_convert=%s',
+                self, to_convert
+            )
+        return to_convert
+
+    @property
+    def proxito_urlconf(self):
+        """
+        Returns a URLConf class that is dynamically inserted via proxito.
+
+        It is used for doc serving on projects that have their own ``urlconf``.
+        """
+        from readthedocs.projects.views.public import ProjectDownloadMedia
+        from readthedocs.proxito.views.serve import ServeDocs
+        from readthedocs.proxito.views.utils import proxito_404_page_handler
+        from readthedocs.proxito.urls import core_urls
+
+        class ProxitoURLConf:
+
+            """A URLConf dynamically inserted by Proxito."""
+
+            proxied_urls = [
+                re_path(
+                    r'{proxied_api_url}api/v2/'.format(proxied_api_url=self.proxied_api_url),
+                    include('readthedocs.api.v2.proxied_urls'),
+                    name='fake_proxied_api'
```

Comment:
Why do we need to proxy these URLs with `/hello/_/` prefix (following my previous comment)? Why not just use `_/` as we are currently doing?

Reply:
Because we need to be able to support proxied API's at non-root URLs for a users who want to use a proxied subpath `docs.company.com/rtd-hosted-project/_/`